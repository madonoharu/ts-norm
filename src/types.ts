/* eslint-disable @typescript-eslint/no-explicit-any */
import { EntitySchema } from "./entity";

export interface Dictionary<T> {
  [id: string]: T | undefined;
}

export type UnionToIntersection<U> = (
  U extends unknown ? (arg: U) => 0 : never
) extends (arg: infer I) => 0
  ? I
  : never;

export type CircularMark = { __mark: "circular" };

export type NoInfer<T> = [T][T extends any ? 0 : never];

export type EntityId = string | number;

export type SchemaFunction<T = unknown> = (
  value: T,
  parent: unknown,
  key: string
) => string;

type ObjectTypeKeys<T> = {
  [P in keyof T]: T[P] extends object ? P : never;
}[keyof T];

export type EntitySchemaDefinition<T> = {
  [P in keyof T & string]?: Schema<T[P] & object>;
};

export type SchemaObject<T extends object> = ObjectTypeKeys<T> extends never
  ? never
  : {
      [P in keyof T]: T[P] extends object ? Schema<T[P]> : never;
    };

export type SchemaArray<T> = Schema<T & object>[];

export type Schema<T extends object> = T extends (infer I)[]
  ? SchemaArray<I>
  : EntitySchema<T, any, any, any, any> | SchemaObject<T>;

export type AnyEntitySchema = EntitySchema<any, string, any, any, any>;
export type AnySchemaArray = AnySchema[];
export type AnySchemaObject = { [key: string]: AnySchema | undefined };
export type AnySchema = AnyEntitySchema | AnySchemaArray | AnySchemaObject;

export type Entities = Record<string, Dictionary<object>>;

export type EntityIdType<E extends AnyEntitySchema> = E extends EntitySchema<
  any,
  any,
  any,
  any,
  infer IdType
>
  ? IdType
  : never;

type EntityTypeImpl<
  Input extends object,
  Definition extends EntitySchemaDefinition<Input>,
  IdAttribute extends string,
  IdType extends EntityId
> = NormalizedSchemaResult<Input, Definition> & Record<IdAttribute, IdType>;

export type EntityType<E extends AnyEntitySchema> = E extends EntitySchema<
  infer Input,
  any,
  infer Definition,
  infer IdAttribute,
  infer IdType
>
  ? EntityTypeImpl<Input, Definition, IdAttribute, IdType>
  : never;

export type NormalizedSchemaResultArray<
  I,
  S extends AnySchemaArray
> = I extends (infer Item)[] ? NormalizedSchemaResult<Item, S[0]>[] : never;

export type NormalizedSchemaResultObject<I, S extends AnySchemaObject> = {
  [K in keyof I]: K extends keyof S
    ? NormalizedSchemaResult<I[K], Extract<S[K], AnySchema>>
    : I[K];
};

export type NormalizedSchemaResult<I, S extends AnySchema> = I extends object
  ? S extends AnyEntitySchema
    ? EntityIdType<S>
    : S extends AnySchemaArray
    ? NormalizedSchemaResultArray<I, S>
    : S extends AnySchemaObject
    ? NormalizedSchemaResultObject<I, S>
    : never
  : I;

type NormalizedSchemaEntitiesUnionImpl<
  Input extends object,
  Key extends string,
  Definition extends EntitySchemaDefinition<Input>,
  IdAttribute extends string,
  IdType extends EntityId
> =
  | Record<
      Key,
      Dictionary<EntityTypeImpl<Input, Definition, IdAttribute, IdType>>
    >
  | NormalizedSchemaEntitiesUnion<Definition>;

type NormalizedSchemaEntitiesUnion<S extends AnySchema> =
  S extends EntitySchema<
    infer Input,
    infer Key,
    infer Definition,
    infer IdAttribute,
    infer IdType
  >
    ? S extends CircularMark
      ? never
      : NormalizedSchemaEntitiesUnionImpl<
          Input,
          Key,
          Definition,
          IdAttribute,
          IdType
        >
    : S extends AnySchemaArray
    ? NormalizedSchemaEntitiesUnion<S[0]>
    : S extends Record<string, AnySchema>
    ? {
        [P in keyof S]: NormalizedSchemaEntitiesUnion<S[P]>;
      }[keyof S]
    : never;

export type NormalizedSchemaEntities<S extends AnySchema> = AnySchema extends S
  ? never
  : UnionToIntersection<NormalizedSchemaEntitiesUnion<S>>;

export type NormalizedSchema<I, S extends AnySchema> = {
  result: NormalizedSchemaResult<I, S>;
  entities: NormalizedSchemaEntities<S>;
};

export function isObject(v: unknown): v is object {
  return typeof v === "object" && Boolean(v);
}

export function isEntitySchema(v: unknown): v is AnyEntitySchema {
  return v instanceof EntitySchema;
}

export function nonNullable<T>(item: T): item is NonNullable<T> {
  return item !== undefined && item !== null;
}
