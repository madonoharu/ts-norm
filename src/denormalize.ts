import {
  AnyEntitySchema,
  AnySchema,
  AnySchemaArray,
  AnySchemaObject,
  EntityId,
  Entities,
  NoInfer,
  NormalizedSchemaEntities,
  NormalizedSchemaResult,
  Schema,
  isObject,
  nonNullable,
  isEntitySchema,
  isEntityId,
} from "./types";

export class Denormalizer {
  cache: Record<string, object> = {};

  constructor(public entities: Entities) {}

  getEntity(input: EntityId, schema: AnyEntitySchema) {
    return this.entities[schema.key]?.[input];
  }

  visitArray(input: object, schema: AnySchemaArray) {
    const localSchema = schema[0];

    if (Array.isArray(input)) {
      return input.map((value) => this.visit(value, localSchema));
    }

    return input;
  }

  visitObject(input: object, schema: AnySchemaObject) {
    const result: Record<string, unknown> = { ...input };

    const keys = Object.keys(schema);

    keys.forEach((key) => {
      const localSchema = schema[key];
      if (result[key] != null && nonNullable(localSchema)) {
        result[key] = this.visit(result[key], localSchema);
      }
    });

    return result;
  }

  visitEntity(id: EntityId, schema: AnyEntitySchema) {
    const { cache } = this;
    let entity = this.getEntity(id, schema);

    if (entity === undefined) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      entity = schema.fallback(id);
    }

    if (!isObject(entity)) {
      return entity;
    }

    const key = `${schema.key}-${id}`;

    if (!cache[key]) {
      const entityCopy = { ...entity };

      // Need to set this first so that if it is referenced further within the
      // denormalization the reference will already exist.
      cache[key] = entityCopy;
      cache[key] = schema.denormalize(entityCopy, this);
    }

    return cache[key];
  }

  visit(input: unknown, schema: AnySchema): unknown {
    if (isEntitySchema(schema)) {
      if (!isEntityId(input)) {
        return input;
      }

      return this.visitEntity(input, schema);
    }

    if (!isObject(input)) {
      return input;
    }

    return Array.isArray(schema)
      ? this.visitArray(input, schema)
      : this.visitObject(input, schema);
  }
}

type SchemaInputType<S extends AnySchema> = S extends Schema<infer T>
  ? T
  : never;

export function denormalize<S extends AnySchema>(
  input: NormalizedSchemaResult<SchemaInputType<S>, S>,
  schema: S,
  entities: NoInfer<NormalizedSchemaEntities<S>>
): SchemaInputType<S> {
  const visitor = new Denormalizer(entities as Entities);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return visitor.visit(input, schema) as SchemaInputType<S>;
}
