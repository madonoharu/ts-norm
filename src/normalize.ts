import {
  AnyEntitySchema,
  AnySchema,
  AnySchemaArray,
  AnySchemaObject,
  EntityId,
  NestedEntities,
  NormalizedSchema,
  isObject,
  nonNullable,
  isEntitySchema,
} from "./types";

export class Normalizer {
  entities: NestedEntities = {};
  cache = new Map<unknown, EntityId>();

  addEntity(schema: AnyEntitySchema, id: EntityId, processedEntity: object) {
    const obj = (this.entities[schema.key] ||= {});
    obj[id] = processedEntity;
  }

  visitArray(input: object, schema: AnySchemaArray) {
    const localSchema = schema[0];
    return Object.values(input).map((value) => this.visit(value, localSchema));
  }

  visitObject(input: Record<string, unknown>, schema: AnySchemaObject) {
    const result = { ...input };

    const keys = Object.keys(schema) as (keyof typeof schema)[];

    keys.forEach((key) => {
      const localSchema = schema[key];

      if (nonNullable(localSchema)) {
        const value = this.visit(input[key], localSchema);

        if (value === undefined) {
          delete result[key];
        } else {
          result[key] = value;
        }
      }
    });

    return result;
  }

  visitEntity(input: object, schema: AnyEntitySchema): EntityId {
    const cachedId = this.cache.get(input);
    if (cachedId !== undefined) {
      return cachedId;
    }

    return schema.normalize(input, this) as EntityId;
  }

  visit(input: unknown, schema: AnySchema): unknown {
    if (!isObject(input)) {
      return input;
    }

    if (isEntitySchema(schema)) {
      return this.visitEntity(input, schema);
    }

    if (Array.isArray(schema)) {
      return this.visitArray(input, schema);
    }

    return this.visitObject(input as Record<string, unknown>, schema);
  }
}

export function normalize<I extends object, S extends AnySchema>(
  input: I,
  schema: S
): NormalizedSchema<I, S> {
  if (!input || typeof input !== "object") {
    throw new Error(
      `Unexpected input given to normalize. Expected type to be "object", found "${
        input === null ? "null" : typeof input
      }".`
    );
  }

  const visitor = new Normalizer();
  const result = visitor.visit(input, schema);

  return {
    entities: visitor.entities,
    result,
  } as NormalizedSchema<I, S>;
}
