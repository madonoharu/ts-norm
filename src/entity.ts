import { Denormalizer } from "./denormalize";
import { Normalizer } from "./normalize";
import {
  nonNullable,
  isEntityId,
  CircularMark,
  EntityId,
  EntitySchemaDefinition,
} from "./types";

export type DefinitionEntries<S extends EntitySchemaDefinition<object>> = [
  keyof S & string,
  NonNullable<S[keyof S & string]>
][];

export type FallbackFn<
  Input extends object,
  Key extends string,
  Definition extends EntitySchemaDefinition<Input>,
  IdAttribute extends string,
  IdType extends EntityId
> = (
  id: IdType,
  schema: EntitySchema<Input, Key, Definition, IdAttribute, IdType>
) => Input;

export interface EntityOptions<
  Input extends object,
  Key extends string,
  Definition extends EntitySchemaDefinition<Input>,
  IdAttribute extends string,
  IdType extends EntityId
> {
  idAttribute?: IdAttribute;
  generateId?: () => IdType;
  fallbackStrategy?: FallbackFn<Input, Key, Definition, IdAttribute, IdType>;
}

export class EntitySchema<
  Input extends object,
  Key extends string,
  Definition extends EntitySchemaDefinition<Input>,
  IdAttribute extends string,
  IdType extends EntityId
> {
  idAttribute: IdAttribute;
  generateId?: EntityOptions<
    Input,
    Key,
    Definition,
    IdAttribute,
    IdType
  >["generateId"];
  fallbackStrategy?: EntityOptions<
    Input,
    Key,
    Definition,
    IdAttribute,
    IdType
  >["fallbackStrategy"];

  constructor(
    public key: Key,
    public definition: Definition,
    options: EntityOptions<Input, Key, Definition, IdAttribute, IdType> = {}
  ) {
    this.idAttribute = (options.idAttribute || "id") as IdAttribute;
    if (options.generateId) {
      this.generateId = options.generateId;
    }
    if (options.fallbackStrategy) {
      this.fallbackStrategy = options.fallbackStrategy;
    }
  }

  define<D2 extends EntitySchemaDefinition<Input>>(
    definition: D2 & EntitySchemaDefinition<Input>
  ): EntitySchema<Input, Key, Definition & D2, IdAttribute, IdType>;

  define<D2 extends EntitySchemaDefinition<Input>>(
    definition: (
      self: this & CircularMark
    ) => D2 & EntitySchemaDefinition<Input>
  ): EntitySchema<Input, Key, Definition & D2, IdAttribute, IdType>;

  define<D2 extends EntitySchemaDefinition<Input>>(
    definition: D2 | ((self: this & CircularMark) => D2)
  ): EntitySchema<Input, Key, Definition & D2, IdAttribute, IdType> {
    if (typeof definition === "function") {
      Object.assign(this.definition, definition(this as this & CircularMark));
    } else {
      Object.assign(this.definition, definition);
    }

    return this as unknown as EntitySchema<
      Input,
      Key,
      Definition & D2,
      IdAttribute,
      IdType
    >;
  }

  getId(input: Input): EntityId | undefined {
    const id = (input as Record<string, unknown>)[this.idAttribute];

    if (isEntityId(id)) {
      return id;
    }

    return this.generateId?.();
  }

  fallback(id: EntityId) {
    return this.fallbackStrategy?.(id as IdType, this);
  }

  definitionEntries(): DefinitionEntries<Definition> {
    return Object.entries(this.definition).filter((entry) =>
      nonNullable(entry[1])
    ) as DefinitionEntries<Definition>;
  }

  normalize(input: Input, visitor: Normalizer): IdType | undefined {
    const id = this.getId(input);

    if (id === undefined) {
      return;
    }

    visitor.cache.set(input, id);

    const processedEntity: Record<string, unknown> = {
      [this.idAttribute]: id,
      ...input,
    };

    this.definitionEntries().forEach(([key, schema]) => {
      if (
        Object.prototype.hasOwnProperty.call(processedEntity, key) &&
        typeof processedEntity[key] === "object"
      ) {
        processedEntity[key] = visitor.visit(processedEntity[key], schema);
      }
    });

    visitor.addEntity(this, id, processedEntity);

    return id as IdType | undefined;
  }

  denormalize(entity: Record<string, unknown>, visitor: Denormalizer) {
    this.definitionEntries().forEach(([key, schema]) => {
      if (Object.prototype.hasOwnProperty.call(entity, key)) {
        entity[key] = visitor.visit(entity[key], schema);
      }
    });

    return entity;
  }
}
