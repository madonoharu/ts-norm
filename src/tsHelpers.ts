/* eslint-disable @typescript-eslint/no-explicit-any */
import { EntitySchema } from "./entity";

export type UnionToIntersection<U> = (
  U extends unknown ? (arg: U) => 0 : never
) extends (arg: infer I) => 0
  ? I
  : never;

export type NoInfer<T> = [T][T extends any ? 0 : never];

export type EntityId = string | number;

export type SchemaFunction<T = unknown> = (
  value: T,
  parent: unknown,
  key: string
) => string;

type ObjectTypeKeys<T> = {
  [P in keyof T]: T[P] extends object ? P : never;
}[keyof T];

export type EntitySchemaDefinition<T> = {
  [P in keyof T & string]?: T[P] extends object ? Schema<T[P]> : never;
};

export type SchemaObject<T extends object> = ObjectTypeKeys<T> extends never
  ? never
  : {
      [P in keyof T]: T[P] extends object ? Schema<T[P]> : never;
    };

export type SchemaArray<T> = Schema<T & object>[];

export type Schema<T extends object> = T extends (infer I)[]
  ? SchemaArray<I>
  : EntitySchema<T, any, any, any, any> | SchemaObject<T>;

export type AnyEntitySchema = EntitySchema<any, string, any, any, any>;
export type AnySchemaArray = AnySchema[];
export type AnySchemaObject = { [key: string]: AnySchema | undefined };
export type AnySchema = AnyEntitySchema | AnySchemaArray | AnySchemaObject;

export type Entities = Record<string, object | undefined>;
export type NestedEntities = Record<string, Entities | undefined>;

export type EntityIdType<E extends AnyEntitySchema> = E extends EntitySchema<
  any,
  any,
  any,
  any,
  infer IdType
>
  ? IdType
  : never;

export type EntityType<E extends AnyEntitySchema> = E extends EntitySchema<
  infer Input,
  any,
  infer Definition,
  infer IdAttribute,
  infer IdType
>
  ? NormalizedSchemaResult<Input, Definition> & Record<IdAttribute, IdType>
  : never;

export type NormalizedSchemaResultArray<
  I,
  S extends AnySchemaArray
> = I extends (infer Item)[] ? NormalizedSchemaResult<Item, S[0]>[] : never;

export type NormalizedSchemaResultObject<I, S extends AnySchemaObject> = {
  [K in keyof I]: NormalizedSchemaResult<
    I[K],
    Extract<S[K & keyof S], AnySchema>
  >;
};

export type NormalizedSchemaResult<I, S extends AnySchema> = I extends object
  ? S extends AnyEntitySchema
    ? EntityIdType<S>
    : S extends AnySchemaArray
    ? NormalizedSchemaResultArray<I, S>
    : S extends AnySchemaObject
    ? NormalizedSchemaResultObject<I, S>
    : never
  : I;

type NormalizedSchemaEntitiesUnionImpl<
  Input extends object,
  Key extends string,
  Definition extends EntitySchemaDefinition<Input>,
  IdAttribute extends string
> =
  | Record<
      Key,
      Record<
        string,
        EntityType<EntitySchema<Input, Key, Definition, IdAttribute, any>>
      >
    >
  | NormalizedSchemaEntitiesUnion<Definition>;

type NormalizedSchemaEntitiesUnion<S extends AnySchema> =
  S extends EntitySchema<
    infer Input,
    infer Key,
    infer Definition,
    infer IdAttribute,
    any
  >
    ? NormalizedSchemaEntitiesUnionImpl<Input, Key, Definition, IdAttribute>
    : S extends AnySchemaArray
    ? NormalizedSchemaEntitiesUnion<S[0]>
    : S extends Record<string, AnySchema>
    ? {
        [P in keyof S]: NormalizedSchemaEntitiesUnion<S[P]>;
      }[keyof S]
    : never;

export type NormalizedSchemaEntities<S extends AnySchema> = AnySchema extends S
  ? never
  : UnionToIntersection<NormalizedSchemaEntitiesUnion<S>>;

export type NormalizedSchema<I, S extends AnySchema> = {
  result: NormalizedSchemaResult<I, S>;
  entities: NormalizedSchemaEntities<S>;
};

export function isObject(v: unknown): v is object {
  return typeof v === "object" && Boolean(v);
}

export function nonNullable<T>(item: T): item is NonNullable<T> {
  return item !== undefined && item !== null;
}

export default function isPlainObject(value: unknown): value is object {
  if (typeof value !== "object" || value === null) return false;

  let proto = value;
  while (Object.getPrototypeOf(proto) !== null) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    proto = Object.getPrototypeOf(proto);
  }

  return Object.getPrototypeOf(value) === proto;
}
