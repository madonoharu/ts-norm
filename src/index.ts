import { EntityOptions, EntitySchema } from "./entity";
import { AnyEntitySchema, EntityId, EntitySchemaDefinition } from "./tsHelpers";

type ExtractIdType<
  Input extends object,
  IdAttribute extends string
> = IdAttribute extends keyof Input
  ? Extract<Input[IdAttribute], EntityId>
  : EntityId;

interface SchemaBuilder<Input extends object> {
  entity<Key extends string>(
    key: Key
    /* eslint-disable @typescript-eslint/ban-types */
  ): EntitySchema<Input, Key, {}, "id", ExtractIdType<Input, "id">>;
  entity<Key extends string, Definition extends EntitySchemaDefinition<Input>>(
    key: Key,
    definition: Definition
  ): EntitySchema<Input, Key, Definition, "id", ExtractIdType<Input, "id">>;
  entity<
    Key extends string,
    Definition extends EntitySchemaDefinition<Input>,
    IdAttribute extends string = "id",
    IdType extends EntityId = ExtractIdType<Input, IdAttribute>
  >(
    key: Key,
    definition: Definition,
    options: EntityOptions<Input, Key, Definition, IdAttribute, IdType>
  ): EntitySchema<Input, Key, Definition, IdAttribute, IdType>;
}

const entity = (
  key: string,
  definition: EntitySchemaDefinition<object> = {},
  options = {}
): AnyEntitySchema => {
  return new EntitySchema(key, definition, options);
};

const builder = {
  entity,
};

export function schema<T extends object>(): SchemaBuilder<T> {
  return builder;
}

export * from "./tsHelpers";
export { normalize } from "./normalize";
export { denormalize } from "./denormalize";
export { EntitySchema } from "./entity";
