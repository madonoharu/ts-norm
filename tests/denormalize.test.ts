/* eslint-disable @typescript-eslint/ban-ts-comment */
import { denormalize, schema } from "../src";
import { expectExactType } from "./utils";

describe("denormalize", () => {
  test("cannot denormalize without a schema", () => {
    expect(() =>
      // @ts-expect-error
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      denormalize({}, undefined, {})
    ).toThrow();
  });

  test("denormalizes entities", () => {
    type Taco = { id: number; type: string };

    const mySchema = schema<Taco>().entity("tacos");

    const entities = {
      tacos: {
        1: { id: 1, type: "foo" },
        2: { id: 2, type: "bar" },
      },
    };

    const result = denormalize([1, 2], [mySchema], entities);

    type Expected = Taco[];
    expectExactType<Expected>()(result).toMatchSnapshot();
  });

  test("denormalizes nested entities", () => {
    type User = {
      id: string;
      name: string;
    };

    type Comment = {
      id: string;
      comment: string;
      user: User;
    };

    type Article = {
      id: string;
      author: User;
      body: string;
      comments: Comment[];
      title: string;
    };

    const user = schema<User>().entity("users");
    const comment = schema<Comment>().entity("comments", {
      user,
    });
    const article = schema<Article>().entity("articles", {
      author: user,
      comments: [comment],
    });

    const entities = {
      articles: {
        123: {
          author: "8472",
          body: "This article is great.",
          comments: ["comment-123-4738"],
          id: "123",
          title: "A Great Article",
        },
      },
      comments: {
        "comment-123-4738": {
          comment: "I like it!",
          id: "comment-123-4738",
          user: "10293",
        },
      },
      users: {
        10293: {
          id: "10293",
          name: "Jane",
        },
        8472: {
          id: "8472",
          name: "Paul",
        },
      },
    };

    type Expected = Article;

    expectExactType<Expected>()(
      denormalize("123", article, entities)
    ).toMatchSnapshot();
  });

  test("set to undefined if schema key is not in entities", () => {
    type Comment = {
      id: string;
      user: object;
    };

    type Article = {
      id: string;
      author: object;
      comments: Comment[];
    };

    const user = schema().entity("users");
    const comment = schema<Comment>().entity("comments", {
      user,
    });
    const article = schema<Article>().entity("articles", {
      author: user,
      comments: [comment],
    });

    const entities = {
      users: {},
      articles: {
        123: {
          id: "123",
          author: "8472",
          comments: ["1"],
        },
      },
      comments: {
        1: {
          id: "1",
          user: "123",
        },
      },
    };

    type Expected = Article;

    expectExactType<Expected>()(
      denormalize("123", article, entities)
    ).toMatchSnapshot();
  });

  test("does not modify the original entities", () => {
    type User = {
      id: string;
      name: string;
    };
    type Article = {
      id: string;
      title: string;
      author: User;
    };

    const user = schema<User>().entity("users");
    const article = schema<Article>().entity("articles", { author: user });

    const entities = Object.freeze({
      articles: Object.freeze({
        123: Object.freeze({
          id: "123",
          title: "A Great Article",
          author: "8472",
        }),
      }),
      users: Object.freeze({
        8472: Object.freeze({
          id: "8472",
          name: "Paul",
        }),
      }),
    });

    expect(() => denormalize("123", article, entities)).not.toThrow();
  });
});
