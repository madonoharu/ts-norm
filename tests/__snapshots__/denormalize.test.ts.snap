// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`denormalize denormalizes entities 1`] = `
Array [
  Object {
    "id": 1,
    "type": "foo",
  },
  Object {
    "id": 2,
    "type": "bar",
  },
]
`;

exports[`denormalize denormalizes nested entities 1`] = `
Object {
  "author": Object {
    "id": "8472",
    "name": "Paul",
  },
  "body": "This article is great.",
  "comments": Array [
    Object {
      "comment": "I like it!",
      "id": "comment-123-4738",
      "user": Object {
        "id": "10293",
        "name": "Jane",
      },
    },
  ],
  "id": "123",
  "title": "A Great Article",
}
`;

exports[`denormalize set to undefined if schema key is not in entities 1`] = `
Object {
  "author": undefined,
  "comments": Array [
    Object {
      "id": "1",
      "user": undefined,
    },
  ],
  "id": "123",
}
`;
